From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: willies952002 <admin@domnian.com>
Date: Fri, 16 Aug 2019 22:18:35 -0400
Subject: [PATCH] Fix Bukkit.createInventory() with type LECTERN

This fixes an issue with Bukkit which makes it possible to open a Lectern interface, but not be able to interact with it (e.g.: change pages). The following changes had to be made:

nms.TileEntityLectern:
- Add `virtual` flag, this is used to stop calls that would attempt to update a block which we do not have, as well as used in the following change.

nms.TileEntityLectern$LecternInventory
- in `a(EntityHuman)`, add `(TileEntityLectern.this.virtual && TileEntityLectern.this.hasBook()) ||` to short-circuit the "can use" logic
- Add `getLectern()` method for use in the following change.

obc.e.CraftHumanEntity#openInventory(Inventory):
- Check if the wrapped inventory is a TileEntityLectern.LecternInventory, and get the Lectern from that

obc.i.u.CraftTileInventoryConverter$Lectern:
- Mark the created lectern as "virtual"
- Override `getInventory(IInventory)` to return a CraftInventoryLectern.

This patch is licensed under the MIT License.
License: https://opensource.org/licenses/MIT

diff --git a/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
index fb7ccb3fcf6e84aa381c2fa2f8d1f7ded8eea311..2ed3465c0d05497ef46e860eb14b9266946ad555 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/LecternBlockEntity.java
@@ -108,7 +108,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
 
         @Override
         public ItemStack removeItem(int slot, int amount) {
-            if (slot == 0) {
+            if (slot == 0 && !LecternBlockEntity.this.virtual) { // PaperDragon
                 ItemStack itemstack = LecternBlockEntity.this.book.split(amount);
 
                 if (LecternBlockEntity.this.book.isEmpty()) {
@@ -123,7 +123,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
 
         @Override
         public ItemStack removeItemNoUpdate(int slot) {
-            if (slot == 0) {
+            if (slot == 0 && !LecternBlockEntity.this.virtual) { // PaperDragon
                 ItemStack itemstack = LecternBlockEntity.this.book;
 
                 LecternBlockEntity.this.book = ItemStack.EMPTY;
@@ -158,7 +158,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
 
         @Override
         public boolean stillValid(Player player) {
-            return LecternBlockEntity.this.level.getBlockEntity(LecternBlockEntity.this.worldPosition) != LecternBlockEntity.this ? false : (player.distanceToSqr((double) LecternBlockEntity.this.worldPosition.getX() + 0.5D, (double) LecternBlockEntity.this.worldPosition.getY() + 0.5D, (double) LecternBlockEntity.this.worldPosition.getZ() + 0.5D) > 64.0D ? false : LecternBlockEntity.this.hasBook());
+            return LecternBlockEntity.this.virtual || (LecternBlockEntity.this.level.getBlockEntity(LecternBlockEntity.this.worldPosition) != LecternBlockEntity.this ? false : (player.distanceToSqr((double) LecternBlockEntity.this.worldPosition.getX() + 0.5D, (double) LecternBlockEntity.this.worldPosition.getY() + 0.5D, (double) LecternBlockEntity.this.worldPosition.getZ() + 0.5D) > 64.0D ? false : LecternBlockEntity.this.hasBook())); // PaperDragon
         }
 
         @Override
@@ -191,6 +191,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
     ItemStack book;
     int page;
     private int pageCount;
+    public boolean virtual = false; // PaperDragon
 
     public LecternBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.LECTERN, pos, state);
@@ -210,6 +211,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
     }
 
     void onBookItemRemove() {
+        if (this.virtual) return; // PaperDragon
         this.page = 0;
         this.pageCount = 0;
         LecternBlock.resetBookState(this.getLevel(), this.getBlockPos(), this.getBlockState(), false);
@@ -227,6 +229,7 @@ public class LecternBlockEntity extends BlockEntity implements Clearable, MenuPr
 
         if (j != this.page) {
             this.page = j;
+            if (this.virtual) return; // PaperDragon
             this.setChanged();
             if (this.level != null) LecternBlock.signalPageChange(this.getLevel(), this.getBlockPos(), this.getBlockState()); // CraftBukkit
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 7c7d05852dd463331110d1dcb71b4d4f5312900f..ed2c67312c8673e48b8b64c60e27960f38aa4313 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -308,6 +308,11 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
             if (craft.getInventory() instanceof MenuProvider) {
                 iinventory = (MenuProvider) craft.getInventory();
             }
+            // PaperDragon start
+            if (craft.getInventory() instanceof net.minecraft.world.level.block.entity.LecternBlockEntity.LecternInventory) {
+                iinventory = ((net.minecraft.world.level.block.entity.LecternBlockEntity.LecternInventory)craft.getInventory()).getLectern();
+            }
+            // PaperDragon end
         }
 
         if (iinventory instanceof MenuProvider) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftTileInventoryConverter.java b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftTileInventoryConverter.java
index 7a7f3f53aef601f124d474d9890e23d87dd96900..bdb2207aaf0a0fbeae04c5303407ec713452958f 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftTileInventoryConverter.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftTileInventoryConverter.java
@@ -158,8 +158,19 @@ public abstract class CraftTileInventoryConverter implements CraftInventoryCreat
 
         @Override
         public Container getTileEntity() {
-            return new LecternBlockEntity(BlockPos.ZERO, Blocks.LECTERN.defaultBlockState()).bookAccess;
+            // PaperDragon start
+            LecternBlockEntity lectern = new LecternBlockEntity(BlockPos.ZERO, Blocks.LECTERN.defaultBlockState());
+            lectern.virtual = true;
+            return lectern.bookAccess;
+            // PaperDragon end
         }
+
+        // PaperDragon start
+        @Override
+        public Inventory getInventory(Container tileEntity) {
+            return new org.bukkit.craftbukkit.inventory.CraftInventoryLectern(tileEntity);
+        }
+        // PaperDragon end
     }
 
     public static class Smoker extends CraftTileInventoryConverter {
